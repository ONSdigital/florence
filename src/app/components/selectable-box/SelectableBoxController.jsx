import React, { Component } from "react";
import PropTypes from "prop-types";

import SelectableBoxItem from "./SelectableBoxItem";

const propTypes = {
    heading: PropTypes.string.isRequired,
    items: PropTypes.arrayOf(PropTypes.object),
    activeItem: PropTypes.object,
    handleItemClick: PropTypes.func.isRequired,
    isUpdating: PropTypes.bool,
};
export default class SelectableBoxController extends Component {
    bindItemClick = itemProps => {
        this.props.handleItemClick(itemProps);
    };

    renderList() {
        return (
            <ul className="selectable-box__list">
                {this.props.items.map((item, index) => {
                    return (
                        <SelectableBoxItem
                            key={index}
                            {...item}
                            isSelected={this.props.activeItem && item.id === this.props.activeItem.id}
                            handleClick={this.bindItemClick}
                        />
                    );
                })}
            </ul>
        );
    }

    render() {
        return (
            <div className="selectable-box">
                <h2 className="selectable-box__heading">
                    {this.props.heading}
                    {this.props.isUpdating && <span className="selectable-box__status loader" />}
                </h2>
                {this.props.items.length > 0 ? (
                    this.renderList()
                ) : (
                    <p className="margin-top--1 margin-right--1 margin-bottom--1 margin-left--1">No items to display</p>
                )}
            </div>
        );
    }
}

SelectableBoxController.propTypes = propTypes;
