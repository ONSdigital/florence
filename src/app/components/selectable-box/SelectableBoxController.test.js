import React from "react";
import SelectableBoxController from "./SelectableBoxController";
import renderer from "react-test-renderer";
import { shallow, mount } from "enzyme";

const items = [
    {
        name: "Item 1",
        id: "1",
    },
    {
        name: "Item 2",
        id: "2",
    },
    {
        name: "Item 3",
        id: "3",
    },
];
const fnMock = jest.fn();

describe("SelectableBoxController", () => {
    describe("when items are passed ", () => {
        it("matches snapshot", () => {
            const props = {
                heading: "Selectable box test",
                items,
                handleItemClick: fnMock,
            };
            const component = renderer.create(<SelectableBoxController {...props} />);

            expect(component.toJSON()).toMatchSnapshot();
        });
    });
    describe("when active item is passed ", () => {
        it("matches snapshot", () => {
            const activeItem = {
                name: "Item 3",
                id: "3",
            };
            const props = {
                heading: "Selectable box test - item selected",
                items,
                activeItem,
                handleItemClick: fnMock,
            };
            const component = renderer.create(<SelectableBoxController {...props} />);

            expect(component.toJSON()).toMatchSnapshot();
        });
    });

    it("fires function from props to handle click on item", () => {
        let itemHasBeenClicked = false;
        const props = {
            heading: "Selectable box test - handle item click",
            items,
            handleItemClick: function (event) {
                itemHasBeenClicked = true;
            },
        };
        const component = mount(<SelectableBoxController {...props} />);
        expect(itemHasBeenClicked).toBe(false);
        const SelectableBoxItem = component.find({ name: "Item 2" });
        SelectableBoxItem.simulate("click");
        expect(itemHasBeenClicked).toBe(true);
    });

    describe("when there are not items", () => {
        it("displays message", () => {
            const props = {
                heading: "",
                items: [],
                handleItemClick: fnMock,
            };
            const component = mount(<SelectableBoxController {...props} />);
            expect(component.text()).toBe("No items to display");
        });
    });
});
