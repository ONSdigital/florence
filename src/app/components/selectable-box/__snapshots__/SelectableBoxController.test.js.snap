// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectableBoxController when active item is passed  matches snapshot 1`] = `
<div
  className="selectable-box"
>
  <h2
    className="selectable-box__heading"
  >
    Selectable box test - item selected
  </h2>
  <ul
    className="selectable-box__list"
  >
    <li
      className="selectable-box__item "
      id="1"
      onClick={[Function]}
    >
      Item 1
    </li>
    <li
      className="selectable-box__item "
      id="2"
      onClick={[Function]}
    >
      Item 2
    </li>
    <li
      className="selectable-box__item selected"
      id="3"
      onClick={[Function]}
    >
      Item 3
    </li>
  </ul>
</div>
`;

exports[`SelectableBoxController when items are passed  matches snapshot 1`] = `
<div
  className="selectable-box"
>
  <h2
    className="selectable-box__heading"
  >
    Selectable box test
  </h2>
  <ul
    className="selectable-box__list"
  >
    <li
      className="selectable-box__item "
      id="1"
      onClick={[Function]}
    >
      Item 1
    </li>
    <li
      className="selectable-box__item "
      id="2"
      onClick={[Function]}
    >
      Item 2
    </li>
    <li
      className="selectable-box__item "
      id="3"
      onClick={[Function]}
    >
      Item 3
    </li>
  </ul>
</div>
`;
